# Netflix Movies Dataset — End-to-End Data Analysis Project

## Project Objective

This project analyzes a large Netflix movie dataset to understand what makes movies financially successful and well received by audiences. The goal is to simulate a real-world analyst workflow starting from raw, messy data and ending with business insights and investment recommendations.

The analysis covers the complete data analysis process including:

- Initial data exploration
- Data cleaning and preprocessing
- Handling missing and inconsistent values
- Feature engineering
- Exploratory Data Analysis (EDA)
- Financial and genre analysis
- Trend analysis across years and languages
- Business interpretation and reporting

The project focuses not only on writing pandas code, but also on developing analytical thinking — understanding what the data suggests, identifying limitations, and translating findings into business decisions.

## Key Business Questions

Some of the main questions explored in this project include:

- Which movie genres generate the highest return on investment (ROI)?
- Do larger budgets guarantee higher revenue?
- Which non-English languages show strong commercial potential?
- What patterns separate successful movies from financial flops?
- How has movie quality and popularity changed over time?
- What type of movie would be considered a safer investment strategy?

## Tools Used

- Python
- Pandas
- NumPy
- Jupyter Notebook

## Dataset Overview

The dataset contains approximately 16,000 Netflix movie records with information such as:

- Title
- Director
- Cast
- Genres
- Language
- Budget
- Revenue
- Ratings
- Popularity
- Release Year
- Date Added to Netflix

The dataset also contains several real-world data quality issues such as:
- Missing values
- Incomplete financial data
- Inconsistent text formatting
- Unusable columns
- Disguised missing values represented as zeros

These issues are identified and handled systematically throughout the project.

In [4]:
import pandas as pd
df = pd.read_csv('data/netflix_raw_data.csv')

In [5]:
df.shape

(16000, 18)

In [6]:
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,genres,language,description,popularity,vote_count,vote_average,budget,revenue
0,10192,Movie,Shrek Forever After,Mike Mitchell,"Mike Myers, Eddie Murphy, Cameron Diaz, Antoni...",United States of America,2010-05-16,2010,6.380,NaN,"Comedy, Adventure, Fantasy, Animation, Family",en,A bored and domesticated Shrek pacts with deal...,203.893,7449,6.380,165000000,752600867
1,27205,Movie,Inception,Christopher Nolan,"Leonardo DiCaprio, Joseph Gordon-Levitt, Ken W...","United Kingdom, United States of America",2010-07-15,2010,8.369,NaN,"Action, Science Fiction, Adventure",en,"Cobb, a skilled thief who commits corporate es...",156.242,37119,8.369,160000000,839030630
2,12444,Movie,Harry Potter and the Deathly Hallows: Part 1,David Yates,"Daniel Radcliffe, Emma Watson, Rupert Grint, T...","United Kingdom, United States of America",2010-11-17,2010,7.744,NaN,"Adventure, Fantasy",en,"Harry, Ron and Hermione walk away from their l...",121.191,19327,7.744,250000000,954305868
3,38757,Movie,Tangled,"Byron Howard, Nathan Greno","Mandy Moore, Zachary Levi, Donna Murphy, Ron P...",United States of America,2010-11-24,2010,7.600,NaN,"Animation, Family, Adventure",en,"Feisty teenager Rapunzel, who has long and mag...",111.762,11638,7.600,260000000,592461732
4,10191,Movie,How to Train Your Dragon,"Chris Sanders, Dean DeBlois","Jay Baruchel, Gerard Butler, Craig Ferguson, A...",United States of America,2010-03-18,2010,7.800,NaN,"Fantasy, Adventure, Animation, Family",en,As the son of a Viking leader on the cusp of m...,110.044,13259,7.800,165000000,494879471


In [7]:
df.tail()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,genres,language,description,popularity,vote_count,vote_average,budget,revenue
15995,1440286,Movie,Festival de Viña del Mar 2025: Ha*Ash,NaN,Ha*ash,Chile,2025-02-24,2025,0.0,NaN,Music,es,NaN,4.931,0,0.0,0,0
15996,1271724,Movie,Man and Woman,Vladimir Kott,"Anna Kotova, Stepan Devonin, Pavel Derevyanko,...",Russia,2025-03-13,2025,0.0,NaN,Drama,ru,Poignant stories about men and women who have ...,4.930,0,0.0,0,0
15997,1426364,Movie,Night of the Dead Sorority Babes,"Angel Nichole Bradford, Steve Hermann","Jessa Flux, Lynn Lowry, Angel Nichole Bradford...",NaN,2025-01-28,2025,1.0,NaN,Horror,en,Two villainous entities initiate gorgeous soro...,4.922,1,1.0,0,0
15998,1411248,Movie,A Dunces Burden,Daniel Kowal,"Riley G, Mitchel Corrado",NaN,2025-03-10,2025,0.0,NaN,NaN,en,A Dunces Burden,4.921,0,0.0,0,0
15999,1405742,Movie,Portal to Hell,Woody Bess,"Trey Holland, Keith David, Richard Kind, Micha...",United States of America,2025-02-22,2025,0.0,NaN,"Comedy, Horror",en,After a debt-collector finds a portal to hell ...,4.921,0,0.0,0,0


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16000 entries, 0 to 15999
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   show_id       16000 non-null  int64  
 1   type          16000 non-null  object 
 2   title         16000 non-null  object 
 3   director      15868 non-null  object 
 4   cast          15796 non-null  object 
 5   country       15534 non-null  object 
 6   date_added    16000 non-null  object 
 7   release_year  16000 non-null  int64  
 8   rating        16000 non-null  float64
 9   duration      0 non-null      float64
 10  genres        15893 non-null  object 
 11  language      16000 non-null  object 
 12  description   15868 non-null  object 
 13  popularity    16000 non-null  float64
 14  vote_count    16000 non-null  int64  
 15  vote_average  16000 non-null  float64
 16  budget        16000 non-null  int64  
 17  revenue       16000 non-null  int64  
dtypes: float64(4), int64(5), o

In [9]:
print("=== NULL VALUES ===")
print(df.isnull().sum())

print("\n=== ZERO VALUES IN NUMERIC COLUMNS ===")
print("budget zeros:", (df['budget'] == 0).sum())
print("revenue zeros:", (df['revenue'] == 0).sum())
print("vote_count zeros:", (df['vote_count'] == 0).sum())
print("vote_average zeros:", (df['vote_average'] == 0).sum())
print("rating zeros:", (df['rating'] == 0).sum())

=== NULL VALUES ===
show_id             0
type                0
title               0
director          132
cast              204
country           466
date_added          0
release_year        0
rating              0
duration        16000
genres            107
language            0
description       132
popularity          0
vote_count          0
vote_average        0
budget              0
revenue             0
dtype: int64

=== ZERO VALUES IN NUMERIC COLUMNS ===
budget zeros: 11153
revenue zeros: 10355
vote_count zeros: 894
vote_average zeros: 899
rating zeros: 899


## Cleaning Plan

### Columns to drop entirely:
- duration: 16,000 nulls, completely empty, no analytical value
- show_id: unique identifier only, no analytical value
- description: text field, not needed for quantitative analysis
- type: dataset is all movies, column has no variation

### Zeros to replace with NaN (disguised missing data):
- budget: 11,153 zeros = 69.7% missing financial data
- revenue: 10,355 zeros = 64.7% missing financial data
  → These will be replaced with NaN so they don't corrupt averages
  → All financial analysis will use a separate financial_df subset

### Zeros to KEEP (not missing data):
- vote_count: 894 zeros = genuinely unrated movies
- vote_average: 899 zeros = genuinely unrated movies  
- rating: 899 zeros = same movies
  → These are real values meaning "no rating exists"
  → Will filter these out only when doing rating analysis

### Real nulls to fill:
- director: 132 nulls → fill with 'Unknown'
- cast: 204 nulls → fill with 'Unknown'  
- country: 466 nulls → fill with 'Unknown'

### Rows to drop:
- genres: 107 null rows → drop entirely
  (only 0.67% of data, genre is essential for analysis)

### Data type fixes:
- date_added: object → datetime
- language: strip whitespace and lowercase for consistency

In [10]:
# Drop useless columns
df.drop(columns=['show_id', 'duration', 'description', 'type'], inplace=True)
print("Columns remaining:", df.columns.tolist())

Columns remaining: ['title', 'director', 'cast', 'country', 'date_added', 'release_year', 'rating', 'genres', 'language', 'popularity', 'vote_count', 'vote_average', 'budget', 'revenue']


In [11]:
# Replace financial zeros with NaN:
import numpy as np

df['budget'] = df['budget'].replace(0, np.nan)
df['revenue'] = df['revenue'].replace(0, np.nan)

# Verify
print("Budget nulls now:", df['budget'].isnull().sum())
print("Revenue nulls now:", df['revenue'].isnull().sum())

Budget nulls now: 11153
Revenue nulls now: 10355


In [12]:
# Fix date type:
df['date_added'] = pd.to_datetime(df['date_added'])

# Verify
print(df['date_added'].dtype)
print(df['date_added'].head())

datetime64[ns]
0   2010-05-16
1   2010-07-15
2   2010-11-17
3   2010-11-24
4   2010-03-18
Name: date_added, dtype: datetime64[ns]


In [13]:
#  Standardize text columns:
df['language'] = df['language'].str.strip().str.lower()
df['country'] = df['country'].str.strip()

# Verify no weird spaces remain
print(df['language'].value_counts().head(5))

language
en    9534
fr    1054
ja     904
ko     876
es     731
Name: count, dtype: int64


In [14]:
# Fill real nulls:
df['director'] = df['director'].fillna('Unknown')
df['cast'] = df['cast'].fillna('Unknown')
df['country'] = df['country'].fillna('Unknown')

# Verify
print("Director nulls:", df['director'].isnull().sum())
print("Cast nulls:", df['cast'].isnull().sum())
print("Country nulls:", df['country'].isnull().sum())

Director nulls: 0
Cast nulls: 0
Country nulls: 0


In [15]:
# Drop genre null rows:
rows_before = len(df)
df = df.dropna(subset=['genres'])
rows_after = len(df)

print(f"Rows dropped: {rows_before - rows_after}")
print(f"Rows remaining: {rows_after}")

Rows dropped: 107
Rows remaining: 15893


In [16]:
# Final verification:
print("=== AFTER CLEANING ===")
print(df.isnull().sum())
print(f"\nFinal shape: {df.shape}")

=== AFTER CLEANING ===
title               0
director            0
cast                0
country             0
date_added          0
release_year        0
rating              0
genres              0
language            0
popularity          0
vote_count          0
vote_average        0
budget          11051
revenue         10248
dtype: int64

Final shape: (15893, 14)


## Phase 1 Complete — Data Cleaning Summary

### What We Started With
- 16,000 rows and 18 columns of raw Netflix movie data
- Multiple data quality issues identified before any cleaning began

### Problems Found and Decisions Made

| Column | Problem | Action | Reason |
|--------|---------|--------|--------|
| show_id | Unique identifier, no analytical value | Dropped | Adds no insight to analysis |
| duration | 16,000 nulls — entirely empty | Dropped | Unusable column |
| description | Free text, not needed for quantitative analysis | Dropped | Not relevant to financial or genre analysis |
| type | All rows are movies — zero variation | Dropped | Single value column adds nothing |
| budget | 11,153 zeros (69.7%) — disguised missing data | Replaced with NaN | Zeros would corrupt all financial averages |
| revenue | 10,355 zeros (64.7%) — disguised missing data | Replaced with NaN | Same reasoning as budget |
| date_added | Stored as object (text) | Converted to datetime64 | Required for all time-based analysis |
| language | Inconsistent spacing and casing | Stripped and lowercased | Ensures consistent grouping |
| director | 132 genuine nulls | Filled with 'Unknown' | Preserves rows, flags missing info clearly |
| cast | 204 genuine nulls | Filled with 'Unknown' | Same reasoning as director |
| country | 466 genuine nulls | Filled with 'Unknown' | Same reasoning as director |
| genres | 107 null rows | Dropped entirely | Genre is essential for core analysis, only 0.67% of data lost |

### Important Distinction — Zeros We Kept
vote_count, vote_average, and rating zeros (894-899 rows) were 
intentionally NOT replaced. These represent genuinely unrated movies 
where no audience rating exists. This is real information, not missing 
data. These rows will be filtered out only when conducting rating-specific 
analysis.

### Final State After Cleaning
- Rows remaining: 15,893 (lost 107 rows — 0.67% of original data)
- Columns remaining: 14 (down from 18)
- All null values resolved except budget (11,051) and revenue (10,248)
  which are intentionally NaN — financial analysis will use a verified
  subset only
- Date column now properly typed for time series analysis
- Text columns standardized for consistent grouping

### Cleaning Quality Assessment
Data loss was minimal at 0.67%. All cleaning decisions were conservative 
— rows were only dropped where the missing column was essential for 
analysis and the volume was small enough to lose safely. The dataset is 
now ready for feature engineering and EDA.

## Phase 2 — Feature Engineering

Raw data rarely contains everything needed for deep analysis. 
Feature engineering creates new meaningful columns derived from 
existing ones — turning raw numbers into analytical tools.

In this phase we create:
- Financial metrics: profit, ROI, flop flag
- Budget categorization: grouping movies by production scale
- Rating quality tiers: translating scores into categories
- Time features: extracting year, month from dates
- Genre explosion: splitting multi-genre strings for genre-level analysis

All financial features are created on a separate subset (financial_df) 
containing only movies with verified budget AND revenue data.

In [17]:
# Create verified financial subset — only movies with real budget and revenue
financial_df = df[df['budget'].notna() & df['revenue'].notna()].copy()

print(f"Full dataset: {len(df):,} movies")
print(f"Financial subset: {len(financial_df):,} movies")
print(f"Coverage: {len(financial_df)/len(df)*100:.1f}% of total library")

Full dataset: 15,893 movies
Financial subset: 3,540 movies
Coverage: 22.3% of total library


In [18]:
# Calculate profit — absolute money made after costs
financial_df['profit'] = financial_df['revenue'] - financial_df['budget']

# Calculate ROI percentage — return relative to investment
financial_df['roi_pct'] = (
    (financial_df['revenue'] - financial_df['budget']) 
    / financial_df['budget'] * 100
)

# Flag flops — movies that failed to recoup their budget
financial_df['was_flop'] = financial_df['revenue'] < financial_df['budget']

# Verify
print(financial_df[['title', 'budget', 'revenue', 'profit', 'roi_pct', 'was_flop']].head())
print(f"\nFlop rate: {financial_df['was_flop'].mean()*100:.1f}%")

                                          title       budget      revenue  \
0                           Shrek Forever After  165000000.0  752600867.0   
1                                     Inception  160000000.0  839030630.0   
2  Harry Potter and the Deathly Hallows: Part 1  250000000.0  954305868.0   
3                                       Tangled  260000000.0  592461732.0   
4                      How to Train Your Dragon  165000000.0  494879471.0   

        profit     roi_pct  was_flop  
0  587600867.0  356.121738     False  
1  679030630.0  424.394144     False  
2  704305868.0  281.722347     False  
3  332461732.0  127.869897     False  
4  329879471.0  199.926952     False  

Flop rate: 36.7%


In [19]:
# Categorize movies by production scale
financial_df['budget_tier'] = pd.cut(
    financial_df['budget'],
    bins=[0, 10_000_000, 50_000_000, 100_000_000, float('inf')],
    labels=['Low (<$10M)', 'Mid ($10-50M)', 'High ($50-100M)', 'Blockbuster ($100M+)']
)

# Check distribution
print(financial_df['budget_tier'].value_counts())
print(financial_df['budget_tier'].value_counts(normalize=True).mul(100).round(1))

budget_tier
Mid ($10-50M)           1457
Low (<$10M)             1404
High ($50-100M)          344
Blockbuster ($100M+)     335
Name: count, dtype: int64
budget_tier
Mid ($10-50M)           41.2
Low (<$10M)             39.7
High ($50-100M)          9.7
Blockbuster ($100M+)     9.5
Name: proportion, dtype: float64


In [20]:
# Categorize vote average into quality tiers
df['rating_tier'] = pd.cut(
    df['vote_average'],
    bins=[0, 5, 6.5, 8, 10],
    labels=['Poor', 'Average', 'Good', 'Excellent']
)

# Check distribution — exclude unrated (vote_average == 0)
rated = df[df['vote_average'] > 0]
print(rated['rating_tier'].value_counts())
print(rated['rating_tier'].value_counts(normalize=True).mul(100).round(1))

rating_tier
Average      7151
Good         6124
Poor         1443
Excellent     356
Name: count, dtype: int64


rating_tier
Average      47.4
Good         40.6
Poor          9.6
Excellent     2.4
Name: proportion, dtype: float64


In [21]:
# Extract time components from date_added
df['year_added'] = df['date_added'].dt.year
df['month_added'] = df['date_added'].dt.month_name()
df['movie_age'] = 2025 - df['release_year']

# Verify
print(df[['title', 'date_added', 'year_added', 'month_added', 'movie_age']].head())

                                          title date_added  year_added  \
0                           Shrek Forever After 2010-05-16        2010   
1                                     Inception 2010-07-15        2010   
2  Harry Potter and the Deathly Hallows: Part 1 2010-11-17        2010   
3                                       Tangled 2010-11-24        2010   
4                      How to Train Your Dragon 2010-03-18        2010   

  month_added  movie_age  
0         May         15  
1        July         15  
2    November         15  
3    November         15  
4       March         15  


In [22]:
# Split multi-genre strings into individual rows for genre-level analysis
df_genres = df.assign(
    genre=df['genres'].str.split(', ')
).explode('genre')

# Check how many unique genres exist
print(f"Unique genres: {df_genres['genre'].nunique()}")
print("\nTop 15 genres by movie count:")
print(df_genres['genre'].value_counts().head(15))

Unique genres: 19

Top 15 genres by movie count:
genre
Drama              6910
Comedy             4533
Thriller           3769
Action             3239
Romance            2571
Horror             2425
Adventure          1768
Crime              1738
Animation          1579
Family             1472
Science Fiction    1454
Fantasy            1450
Mystery            1242
Documentary         916
History             732
Name: count, dtype: int64


In [23]:
# Verify all new columns created correctly
print("=== MAIN DATAFRAME NEW COLUMNS ===")
print(df[['rating_tier', 'year_added', 'month_added', 'movie_age']].head())
print(f"\nMain df shape: {df.shape}")

print("\n=== FINANCIAL DATAFRAME NEW COLUMNS ===")
print(financial_df[['profit', 'roi_pct', 'was_flop', 'budget_tier']].head())
print(f"\nFinancial df shape: {financial_df.shape}")

print("\n=== GENRE DATAFRAME ===")
print(f"Genre df shape: {df_genres.shape}")
print(f"Unique genres: {df_genres['genre'].nunique()}")

=== MAIN DATAFRAME NEW COLUMNS ===
  rating_tier  year_added month_added  movie_age
0     Average        2010         May         15
1   Excellent        2010        July         15
2        Good        2010    November         15
3        Good        2010    November         15
4        Good        2010       March         15

Main df shape: (15893, 18)

=== FINANCIAL DATAFRAME NEW COLUMNS ===
        profit     roi_pct  was_flop           budget_tier
0  587600867.0  356.121738     False  Blockbuster ($100M+)
1  679030630.0  424.394144     False  Blockbuster ($100M+)
2  704305868.0  281.722347     False  Blockbuster ($100M+)
3  332461732.0  127.869897     False  Blockbuster ($100M+)
4  329879471.0  199.926952     False  Blockbuster ($100M+)

Financial df shape: (3540, 18)

=== GENRE DATAFRAME ===
Genre df shape: (37460, 19)
Unique genres: 19


## Phase 2 Complete — Feature Engineering Summary

Starting from 15,893 cleaned movies, we created three working 
dataframes and added 8 new analytical columns that the raw data 
couldn't support.

### Three Working Dataframes
| Dataframe | Rows | Purpose |
|-----------|------|---------|
| df | 15,893 | All general analysis |
| financial_df | 3,540 | Budget, revenue, ROI, flop analysis only |
| df_genres | 37,460 | Genre-level analysis (exploded) |

### Key Observations From Feature Engineering

**Financial Coverage is Limited**
Only 3,540 movies (22.3%) have verified financial data. Nearly 78% 
of the library has no budget or revenue records. All financial 
findings apply to this subset only and should not be generalized 
to the full library.

**Flop Rate is Significant**
36.7% of movies with financial data failed to recoup their budget. 
More than 1 in 3 movies loses money — this is the baseline risk 
any investor needs to understand before entering film production.

**Budget Distribution Skews Small**
80.9% of financially verified movies fall in the Low or Mid budget 
range (under $50M). Blockbuster productions ($100M+) represent 
only 9.5% of the dataset — meaning most Netflix content comes from 
mid-range and independent productions, not Hollywood tentpoles.

**Audience Ratings Cluster in the Middle**
88% of rated movies fall in the Average or Good tier (5.0–8.0). 
Only 2.4% achieve Excellent status (above 8.0), while 9.6% are 
rated Poor. The library is predominantly decent-but-not-outstanding 
content with very few standout films.

**19 Distinct Genres**
Drama dominates with 6,910 appearances, followed by Comedy (4,533) 
and Thriller (3,769). Note these are appearances not unique movies 
since one movie can belong to multiple genres. Genre overlap will 
be important context when interpreting genre analysis in Phase 3.

---
*Moving to Phase 3 — EDA and Business Questions*

## Phase 3 — EDA and Business Questions

With clean data and engineered features in place, we now answer 
8 specific business questions a real investor or content strategist 
would care about. Each question is answered with code, numbers, 
and a written interpretation.

The goal is not just to produce outputs but to extract meaning — 
every finding is followed by what it actually suggests for 
decision making.

Three dataframes in use:
- df — general library analysis
- financial_df — all financial questions
- df_genres — all genre-level analysis

In [24]:
# Question 1 — What does the Netflix library look like overall?

# Release year distribution — last 15 years
print("=== MOVIES BY RELEASE YEAR (Last 15 years) ===")
print(df['release_year'].value_counts().sort_index().tail(15))

# How many movies added to Netflix per year
print("\n=== MOVIES ADDED TO NETFLIX PER YEAR ===")
print(df.groupby('year_added')['title'].count())

# Language breakdown
print("\n=== TOP 10 LANGUAGES ===")
print(df['language'].value_counts().head(10))

=== MOVIES BY RELEASE YEAR (Last 15 years) ===
release_year
2011    1000
2012     996
2013     999
2014     999
2015     999
2016     997
2017     995
2018     998
2019     995
2020     997
2021     997
2022     997
2023     998
2024     993
2025     934
Name: count, dtype: int64

=== MOVIES ADDED TO NETFLIX PER YEAR ===
year_added
2010     999
2011    1000
2012     996
2013     999
2014     999
2015     999
2016     997
2017     995
2018     998
2019     995
2020     997
2021     997
2022     997
2023     998
2024     993
2025     934
Name: title, dtype: int64

=== TOP 10 LANGUAGES ===
language
en    9478
fr    1048
ja     897
ko     864
es     723
zh     373
it     311
hi     287
de     278
ru     190
Name: count, dtype: int64


In [25]:
# Question 2 — Which genres dominate the library vs which make the most money?

# Genre count in full library
print("=== TOP 10 GENRES BY VOLUME ===")
print(df_genres['genre'].value_counts().head(10))

# Genre average ROI — financial subset only
print("\n=== AVERAGE ROI BY GENRE ===")
financial_genres = financial_df.assign(
    genre=financial_df['genres'].str.split(', ')
).explode('genre')

print(
    financial_genres.groupby('genre')['roi_pct']
    .mean()
    .sort_values(ascending=False)
    .head(10)
    .round(1)
)

=== TOP 10 GENRES BY VOLUME ===
genre
Drama        6910
Comedy       4533
Thriller     3769
Action       3239
Romance      2571
Horror       2425
Adventure    1768
Crime        1738
Animation    1579
Family       1472
Name: count, dtype: int64

=== AVERAGE ROI BY GENRE ===
genre
Documentary    7226255.0
Music          2340429.9
History         111830.0
Animation        97571.9
Drama            18033.2
Comedy            1572.6
Horror             745.5
Thriller           488.2
Mystery            361.2
Action             297.9
Name: roi_pct, dtype: float64


In [26]:
# Question 3 — Does bigger budget mean bigger revenue?

# Correlation between budget, revenue and rating
print("=== CORRELATION MATRIX ===")
print(financial_df[['budget', 'revenue', 'vote_average']].corr().round(3))

# Average revenue by budget tier
print("\n=== AVERAGE REVENUE BY BUDGET TIER ===")
print(financial_df.groupby('budget_tier')['revenue'].mean().apply(lambda x: f"${x:,.0f}"))

# Average ROI by budget tier
print("\n=== AVERAGE ROI BY BUDGET TIER ===")
print(financial_df.groupby('budget_tier')['roi_pct'].mean().round(1))

=== CORRELATION MATRIX ===
              budget  revenue  vote_average
budget         1.000    0.748         0.121
revenue        0.748    1.000         0.223
vote_average   0.121    0.223         1.000

=== AVERAGE REVENUE BY BUDGET TIER ===
budget_tier
Low (<$10M)              $15,941,825
Mid ($10-50M)            $60,189,420
High ($50-100M)         $225,532,161
Blockbuster ($100M+)    $536,447,948
Name: revenue, dtype: object

=== AVERAGE ROI BY BUDGET TIER ===
budget_tier
Low (<$10M)             196548.7
Mid ($10-50M)              149.5
High ($50-100M)            201.4
Blockbuster ($100M+)       214.1
Name: roi_pct, dtype: float64


C:\Users\dacha\AppData\Local\Temp\ipykernel_31140\1932641165.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(financial_df.groupby('budget_tier')['revenue'].mean().apply(lambda x: f"${x:,.0f}"))
C:\Users\dacha\AppData\Local\Temp\ipykernel_31140\1932641165.py:13: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(financial_df.groupby('budget_tier')['roi_pct'].mean().round(1))


In [27]:
# Question 4 — What does a flop look like vs a success?

# Overall flop rate
print("=== OVERALL FLOP RATE ===")
print(financial_df['was_flop'].value_counts(normalize=True).mul(100).round(1))

# Flop rate by budget tier
print("\n=== FLOP RATE BY BUDGET TIER ===")
print(financial_df.groupby('budget_tier')['was_flop'].mean().mul(100).round(1))

# Average audience rating — flops vs successes
print("\n=== AVERAGE RATING: FLOPS VS SUCCESSES ===")
print(financial_df.groupby('was_flop')['vote_average'].mean().round(2))

=== OVERALL FLOP RATE ===
was_flop
False    63.3
True     36.7
Name: proportion, dtype: float64

=== FLOP RATE BY BUDGET TIER ===
budget_tier
Low (<$10M)             43.3
Mid ($10-50M)           39.5
High ($50-100M)         23.3
Blockbuster ($100M+)    10.4
Name: was_flop, dtype: float64

=== AVERAGE RATING: FLOPS VS SUCCESSES ===
was_flop
False    6.68
True     6.29
Name: vote_average, dtype: float64


C:\Users\dacha\AppData\Local\Temp\ipykernel_31140\2509425447.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(financial_df.groupby('budget_tier')['was_flop'].mean().mul(100).round(1))


In [28]:
# Question 5 — Which languages produce commercially successful non-English movies?

# Filter non-English with real financial data
non_english = financial_df[financial_df['language'] != 'en']

# Average revenue by language
print("=== AVERAGE REVENUE BY LANGUAGE (Non-English) ===")
print(
    non_english.groupby('language')['revenue']
    .mean()
    .sort_values(ascending=False)
    .head(10)
    .apply(lambda x: f"${x:,.0f}")
)

# Movie count with verified financial data by language
print("\n=== MOVIE COUNT WITH FINANCIAL DATA BY LANGUAGE ===")
print(
    non_english.groupby('language')['title']
    .count()
    .sort_values(ascending=False)
    .head(10)
)

=== AVERAGE REVENUE BY LANGUAGE (Non-English) ===
language
zh    $224,957,380
kn     $75,429,274
cn     $73,315,355
ja     $66,654,869
te     $63,866,329
ko     $41,959,352
ta     $40,436,455
hi     $36,819,947
th     $25,146,496
fa     $24,426,169
Name: revenue, dtype: object

=== MOVIE COUNT WITH FINANCIAL DATA BY LANGUAGE ===
language
fr    203
hi    126
es     89
zh     65
ko     61
ru     54
it     47
ja     36
te     19
da     18
Name: title, dtype: int64


In [29]:
# Question 6 — Has movie quality changed over the years?

# Average rating by release year — last 15 years
print("=== AVERAGE RATING BY RELEASE YEAR ===")
rated = df[df['vote_average'] > 0]
print(rated.groupby('release_year')['vote_average'].mean().tail(15).round(2))

# Average popularity by release year
print("\n=== AVERAGE POPULARITY BY RELEASE YEAR ===")
print(df.groupby('release_year')['popularity'].mean().tail(15).round(2))

=== AVERAGE RATING BY RELEASE YEAR ===
release_year
2011    6.15
2012    6.12
2013    6.20
2014    6.21
2015    6.23
2016    6.23
2017    6.34
2018    6.37
2019    6.46
2020    6.43
2021    6.50
2022    6.41
2023    6.43
2024    6.42
2025    6.46
Name: vote_average, dtype: float64

=== AVERAGE POPULARITY BY RELEASE YEAR ===
release_year
2011    11.91
2012    12.56
2013    13.57
2014    15.66
2015    14.72
2016    15.49
2017    15.83
2018    15.73
2019    15.89
2020    13.30
2021    16.50
2022    19.48
2023    25.74
2024    74.33
2025    35.59
Name: popularity, dtype: float64


In [30]:
# Question 7 — Which directors produce the highest rated content?

# Directors with 3+ movies sorted by average rating
print("=== TOP DIRECTORS BY AVERAGE RATING (min 3 movies) ===")
director_stats = (
    df[df['director'] != 'Unknown']
    .groupby('director')
    .agg(
        movie_count=('title', 'count'),
        avg_rating=('vote_average', 'mean')
    )
    .query('movie_count >= 3')
    .sort_values('avg_rating', ascending=False)
    .head(10)
    .round(2)
)
print(director_stats)

# Genre average rating
print("\n=== AVERAGE RATING BY GENRE ===")
print(
    df_genres[df_genres['vote_average'] > 0]
    .groupby('genre')['vote_average']
    .mean()
    .sort_values(ascending=False)
    .round(2)
)

=== TOP DIRECTORS BY AVERAGE RATING (min 3 movies) ===
                             movie_count  avg_rating
director                                            
Park Jun-soo                           3        8.60
Michael Goguen                         3        8.19
Ian Bonhôte, Peter Ettedgui            3        7.98
Tatsuya Oishi                          3        7.97
Masayuki Kojima                        3        7.93
Daniel Rezende                         3        7.93
Christopher Nolan                      6        7.90
Sam Wrench                             3        7.90
Masamitsu Hidaka                       3        7.87
Alex Kendrick                          4        7.85

=== AVERAGE RATING BY GENRE ===
genre
Documentary        7.10
Music              6.96
Animation          6.94
History            6.77
Family             6.68
War                6.63
Fantasy            6.48
Drama              6.46
Adventure          6.46
Crime              6.35
Romance            6.33
Comedy

In [31]:
# Question 8 — Investment recommendation

# Safe bet profile — Horror and Thriller ROI vs flop rate
print("=== HORROR AND THRILLER PROFILE ===")
horror_thriller = financial_genres[
    financial_genres['genre'].isin(['Horror', 'Thriller'])
]
print(
    horror_thriller.groupby('genre').agg(
        avg_roi=('roi_pct', 'mean'),
        avg_budget=('budget', 'mean'),
        flop_rate=('was_flop', 'mean'),
        movie_count=('title', 'count')
    ).round(2)
)

# Best performing non-English languages by ROI
print("\n=== NON-ENGLISH LANGUAGES BY AVERAGE ROI ===")
print(
    non_english.groupby('language')['roi_pct']
    .mean()
    .sort_values(ascending=False)
    .head(8)
    .round(1)
)

=== HORROR AND THRILLER PROFILE ===
          avg_roi   avg_budget  flop_rate  movie_count
genre                                                 
Horror     745.50  14731373.86       0.35          463
Thriller   488.23  27756389.94       0.42          939

=== NON-ENGLISH LANGUAGES BY AVERAGE ROI ===
language
fr    137579.6
mr      2650.0
zh      2408.2
ja      2309.2
kn      1810.5
th      1281.7
hi       735.6
ko       710.5
Name: roi_pct, dtype: float64


## Phase 3 Complete — EDA and Business Questions

### Q1 — Netflix Library Overview
The dataset is remarkably evenly distributed — approximately 1,000 
movies per year from 2010 through 2024, with 2025 slightly lower at 
934 likely because the year is incomplete. This uniform distribution 
suggests the dataset was constructed with balanced sampling rather 
than reflecting actual Netflix addition patterns. English dominates 
the library at 9,534 movies, followed by French (1,054), Japanese 
(904), and Korean (876) — confirming Netflix's strong non-English 
content investment particularly in Asian markets.

### Q2 — Genre Volume vs Genre Profitability
There is a clear disconnect between what Netflix has most of and 
what actually makes money. Drama leads the library by volume (6,910 
appearances) but sits mid-table on ROI. Documentary and Music show 
astronomically high ROI figures — however these numbers are 
unreliable due to near-zero budgets recorded in the dataset making 
percentage returns mathematically extreme. Among genres with 
credible budget data, Horror (745% average ROI) and Thriller (488%) 
are the standout performers. The implication is that Netflix 
prioritizes Drama and Comedy for audience volume while Horror and 
Thriller deliver the strongest financial returns per dollar invested.

### Q3 — Budget vs Revenue
Budget and revenue show a strong positive correlation of 0.748 — 
higher spending generally produces higher revenue. However the ROI 
story tells the opposite: Low budget films (<$10M) show an average 
ROI of 196,549% driven by extreme outliers, while Mid, High, and 
Blockbuster tiers cluster between 149-214%. This means big budget 
films generate more absolute revenue but are not meaningfully more 
efficient investments than smaller productions. Spending more buys 
bigger gross — it does not buy proportionally better returns.

### Q4 — Flops vs Successes
36.7% of movies with financial data failed to recoup their budget 
— more than 1 in 3 films loses money. Flop risk drops dramatically 
as budget increases: Low budget films flop at 43.3%, Mid at 39.5%, 
High at 23.3%, and Blockbusters at only 10.4%. However audience 
ratings tell a nuanced story — flops average 6.29 vs successes at 
6.68, a difference of only 0.39 points. Box office failure and 
audience disappointment are not the same thing. A film can be 
genuinely well-received by audiences and still lose money, 
suggesting marketing, distribution, and timing are as important 
as quality.

### Q5 — Non-English Commercial Potential
Chinese (zh) leads non-English revenue by a significant margin at 
$224.9M average, followed by Kannada ($75.4M), Cantonese ($73.3M), 
Japanese ($66.7M), and Telugu ($63.9M). However volume context 
matters — French has the most movies with financial data (203) but 
lower average revenue ($not shown). Chinese and South Indian 
languages (Kannada, Telugu, Tamil) show high average revenue but 
from smaller sample sizes (65, 19, and 19 movies respectively), 
meaning results may be skewed by a few blockbusters. Korean (61 
movies, $41.9M average) represents the most balanced combination 
of volume and commercial performance among non-English markets.

### Q6 — Movie Quality Over Time
Audience ratings show a slow but consistent improvement from 6.15 
in 2011 to a peak of 6.50 in 2021, before settling around 6.41-6.46 
in recent years. The improvement is modest — less than half a point 
over 14 years — suggesting overall quality has been stable rather 
than dramatically improving. Popularity tells a more striking story: 
relatively flat through 2022, then a sharp spike to 74.33 in 2024 
before dropping to 35.59 in 2025. The 2024 spike likely reflects 
a small number of extremely popular titles distorting the average 
rather than a genuine industry-wide trend.

### Q7 — Directors and Genre Quality
Christopher Nolan is the only widely recognized director in the 
top 10 by average rating with 6 movies averaging 7.90 — the rest 
are specialists in animation, documentary, or niche genres with 
only 3 movies each, making their rankings statistically fragile. 
By genre, Documentary leads ratings at 7.10 followed by Music 
(6.96) and Animation (6.94). Horror sits at the bottom at 5.72 
confirming a consistent pattern — Horror generates strong financial 
returns but scores lowest on audience satisfaction. Quality and 
profitability point in opposite directions for Horror content.

### Q8 — Investment Recommendation
Based on the full analysis, the strongest risk-adjusted investment 
profile for a Netflix movie combines a mid-range budget between 
$10-50M, a Horror or Thriller genre, and consideration of Korean 
or Chinese language productions. Horror delivers 745% average ROI 
on an average budget of $14.7M — the lowest average spend among 
analyzed genres — while maintaining a flop rate of 35% which, 
though not low in absolute terms, is significantly better than the 
43% flop rate seen in unstructured low-budget productions. The 
strong budget-revenue correlation of 0.748 confirms that some 
minimum investment is necessary to generate meaningful returns, 
making the sub-$10M tier riskier despite its theoretical ROI 
figures. Korean content offers the best combination of sample 
volume and commercial performance among non-English markets at 
$41.9M average revenue across 61 verified titles. The data 
therefore points toward a mid-budget Korean or English-language 
Horror or Thriller as the most defensible investment strategy — 
high enough budget to ensure quality production, genre with proven 
ROI history, and language market with demonstrated commercial 
appetite.

---

### Phase 3 Data Quality Notes
- Documentary and Music ROI figures (7.2M% and 2.3M%) are 
  statistically unreliable due to near-zero recorded budgets 
  and should not be used for investment decisions
- Non-English revenue averages for languages with fewer than 
  30 movies in the financial subset should be treated as 
  indicative only, not conclusive
- All financial findings apply to the 3,540 movie verified 
  subset — 22.3% of the full library

---
*Moving to Phase 4 — Cleaning Log and Data Handoff Note*

## Phase 4 — Cleaning Log and Data Handoff Note

This final phase documents every transformation made to the raw 
dataset in a format that can be audited, reproduced, or handed 
to another analyst or data scientist. Professional data work 
requires full transparency about what was changed and why.

In [32]:
# Final state of all three working dataframes
print("=== FINAL DATAFRAME SUMMARY ===")
print(f"Main df:        {df.shape[0]:,} rows, {df.shape[1]} columns")
print(f"Financial df:   {financial_df.shape[0]:,} rows, {financial_df.shape[1]} columns")
print(f"Genre df:       {df_genres.shape[0]:,} rows, {df_genres.shape[1]} columns")

print("\n=== REMAINING NULLS IN MAIN DF ===")
print(df.isnull().sum()[df.isnull().sum() > 0])

print("\n=== COLUMN DTYPES ===")
print(df.dtypes)

=== FINAL DATAFRAME SUMMARY ===
Main df:        15,893 rows, 18 columns
Financial df:   3,540 rows, 18 columns
Genre df:       37,460 rows, 19 columns

=== REMAINING NULLS IN MAIN DF ===
budget         11051
revenue        10248
rating_tier      819
dtype: int64

=== COLUMN DTYPES ===
title                   object
director                object
cast                    object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                 float64
genres                  object
language                object
popularity             float64
vote_count               int64
vote_average           float64
budget                 float64
revenue                float64
rating_tier           category
year_added               int32
month_added             object
movie_age                int64
dtype: object


## Phase 4 Complete — Cleaning Log and Data Handoff Note

---

### Cleaning Log

| Column | Original State | Problem | Action Taken | Reason |
|--------|---------------|---------|--------------|--------|
| show_id | int64, no nulls | Identifier only, no analytical value | Dropped | Adds nothing to analysis |
| duration | float64, 16,000 nulls | Entirely empty column | Dropped | Completely unusable |
| description | object, 132 nulls | Free text, not needed | Dropped | Not relevant to quantitative analysis |
| type | object, no nulls | All values identical — "Movie" | Dropped | Zero variation, no analytical value |
| budget | int64, no nulls | 11,153 zeros = disguised missing data | Zeros replaced with NaN | Zeros would corrupt all financial averages |
| revenue | int64, no nulls | 10,355 zeros = disguised missing data | Zeros replaced with NaN | Same reasoning as budget |
| date_added | object, no nulls | Dates stored as text | Converted to datetime64 | Required for all time-based analysis |
| language | object, no nulls | Inconsistent spacing and casing | Stripped and lowercased | Ensures consistent groupby results |
| director | object, 132 nulls | Genuine missing values | Filled with 'Unknown' | Preserves rows, flags missing clearly |
| cast | object, 204 nulls | Genuine missing values | Filled with 'Unknown' | Same reasoning as director |
| country | object, 466 nulls | Genuine missing values | Filled with 'Unknown' | Same reasoning as director |
| genres | object, 107 nulls | Rows unusable without genre | 107 rows dropped | Genre essential for core analysis |
| vote_count | int64, 894 zeros | Genuinely unrated movies | Kept as-is | Zeros are real data not missing data |
| vote_average | float64, 899 zeros | Genuinely unrated movies | Kept as-is | Same reasoning as vote_count |
| rating | float64, 899 zeros | Genuinely unrated movies | Kept as-is | Same reasoning as vote_count |

---

### Features Engineered

| Column | Created In | Formula / Method | Purpose |
|--------|-----------|-----------------|---------|
| profit | financial_df | revenue - budget | Absolute earnings after cost |
| roi_pct | financial_df | (revenue - budget) / budget × 100 | Return efficiency per dollar spent |
| was_flop | financial_df | revenue < budget | Boolean risk flag |
| budget_tier | financial_df | pd.cut into 4 bins | Group by production scale |
| rating_tier | df | pd.cut into 4 bins | Translate scores to categories |
| year_added | df | date_added.dt.year | Enable yearly trend analysis |
| month_added | df | date_added.dt.month_name() | Enable seasonal pattern analysis |
| movie_age | df | 2025 - release_year | Age of film at time of analysis |
| genre (exploded) | df_genres | str.split + explode | Individual genre-level analysis |

---

### Data Handoff Note

The cleaned dataset contains 15,893 Netflix movies spanning 2010 
to 2025 across 19 genres and 20+ languages. All structural issues 
from the raw data have been resolved — wrong data types corrected, 
text standardized, and genuinely empty columns removed. 

Financial analysis is restricted to a verified subset of 3,540 
movies (22.3% of the library) where both budget and revenue records 
exist. The remaining 77.7% had zeros recorded for financial fields 
which have been converted to NaN. Any model trained on budget or 
revenue data should be built on financial_df only and findings 
should not be generalized to the full library.

Three important caveats for downstream modeling:
1. vote_count, vote_average, and rating contain genuine zeros 
   representing unrated movies — these should be filtered out 
   for any rating-based model rather than treated as nulls
2. The genres column contains multiple values per row — 
   exploding is required before using genre as a model feature
3. Documentary and Music ROI figures are extreme outliers driven 
   by near-zero recorded budgets and should be excluded from 
   any financial modeling or treated as a separate category

The dataset is ready for visualization and machine learning 
with the caveats above clearly understood.

---

## Cleaning Phase Complete

| Phase | Status |
|-------|--------|
| Phase 1 — Data Cleaning | Done |
| Phase 2 — Feature Engineering | Done |
| Phase 3 — EDA and Business Questions | Done |
| Phase 4 — Cleaning Log and Handoff | Done |